examples/virtio-snd/Makefile

```diff
@@ -1,311 +1,60 @@
-							--home $(addprefix $(BUILD_DIR)/, $(CLIENT_VM_USERLEVEL_ELFS))
-
-
-$(SND_DRIVER_VM_DTB): $(SND_DRIVER_VM_BASE_DTS) $(SND_DRIVER_VM_DTS_OVERLAYS)
-	sh ../../tools/dtscat $^ > $(BUILD_DIR)/snd_driver_vm.dts
-	$(DTC) -q -I dts -O dtb $(BUILD_DIR)/snd_driver_vm.dts > $@
-
-$(BUILD_DIR)/snd_driver_images.o: $(LIBVMM_TOOLS)/package_guest_images.S $(SND_DRIVER_VM_LINUX) $(SND_DRIVER_VM_DTB) $(SND_DRIVER_VM_INITRD_OVERLAYED)
-	$(CC) -c -g3 -x assembler-with-cpp \
-					-DGUEST_KERNEL_IMAGE_PATH=\"$(SND_DRIVER_VM_LINUX)\" \
-					-DGUEST_DTB_IMAGE_PATH=\"$(SND_DRIVER_VM_DTB)\" \
-					-DGUEST_INITRD_IMAGE_PATH=\"$(SND_DRIVER_VM_INITRD_OVERLAYED)\" \
-					-target aarch64-none-elf \
-					$< -o $@
-
-$(CLIENT_VM_DTB): $(CLIENT_VM_BASE_DTS) $(CLIENT_VM_DTS_OVERLAYS)
-	sh ../../tools/dtscat $^ > $(BUILD_DIR)/client_vm.dts
-	$(DTC) -q -I dts -O dtb $(BUILD_DIR)/client_vm.dts > $@
-
-$(BUILD_DIR)/client_images.o: $(LIBVMM_TOOLS)/package_guest_images.S $(CLIENT_VM_LINUX) $(CLIENT_VM_DTB) $(CLIENT_VM_INITRD_OVERLAYED)
-	$(CC) -c -g3 -x assembler-with-cpp \
-					-DGUEST_KERNEL_IMAGE_PATH=\"$(CLIENT_VM_LINUX)\" \
-					-DGUEST_DTB_IMAGE_PATH=\"$(CLIENT_VM_DTB)\" \
-					-DGUEST_INITRD_IMAGE_PATH=\"$(CLIENT_VM_INITRD_OVERLAYED)\" \
-					-target aarch64-none-elf \
-					$< -o $@
-
-$(BUILD_DIR)/user_sound/%.o: $(UIO_DRIVERS_DIR)/snd/%.c
-	mkdir -p $(BUILD_DIR)/user_sound
-	$(CC_USERLEVEL) -c $(CFLAGS_USERLEVEL) $^ -o $@
-
-$(BUILD_DIR)/user_sound.elf: $(BUILD_DIR)/user_sound/main.o $(BUILD_DIR)/user_sound/stream.o $(BUILD_DIR)/user_sound/queue.o $(BUILD_DIR)/user_sound/convert.o
-	$(CC_USERLEVEL) $(CFLAGS_USERLEVEL) $^ -o $@
-	patchelf --set-interpreter /lib64/ld-linux-aarch64.so.1 $@
-
-$(BUILD_DIR)/%.elf: $(VM_USERLEVEL_DIR)/%.c
-	$(CC_USERLEVEL) $(CFLAGS_USERLEVEL) $^ -o $@
-	patchelf --set-interpreter /lib64/ld-linux-aarch64.so.1 $@
-
-$(BUILD_DIR)/%.o: %.c Makefile
-	$(CC) -c $(CFLAGS) $< -o $@
-
-$(BUILD_DIR)/%.o: $(LIBVMM_SRC_DIR)/%.c Makefile
-	$(CC) -c $(CFLAGS) $< -o $@
-
-$(BUILD_DIR)/%.o: $(LIBVMM_SRC_DIR)/util/%.c Makefile
-	$(CC) -c $(CFLAGS) $< -o $@
-
-$(BUILD_DIR)/%.o: $(LIBVMM_SRC_DIR)/arch/aarch64/%.c Makefile
-	$(CC) -c $(CFLAGS) $< -o $@
-
-$(BUILD_DIR)/%.o: $(LIBVMM_SRC_DIR)/arch/aarch64/vgic/%.c Makefile
-	$(CC) -c $(CFLAGS) $< -o $@
-
-$(BUILD_DIR)/%.o: $(LIBVMM_SRC_DIR)/virtio/%.c Makefile
-	$(CC) -c $(CFLAGS) $< -o $@
+export BUILD_DIR:=$(abspath $(BUILD_DIR))
+export MICROKIT_SDK:=$(abspath $(MICROKIT_SDK))
+export VIRTIO_EXAMPLE:=$(abspath .)
 
-$(BUILD_DIR)/sound_%.o: $(SDDF_SOUND_COMPONENTS)/%.c Makefile
-	$(CC) -c $(CFLAGS) $< -o $@
+export TARGET := aarch64-none-elf
+export CC := clang
+export CC_USERLEVEL := zig cc
+export LD := ld.lld
+export AS := llvm-as
+export AR := llvm-ar
+export DTC := dtc
+export RANLIB := llvm-ranlib
+export MICROKIT_TOOL ?= $(MICROKIT_SDK)/bin/microkit
+export SDDF=$(abspath ../../dep/sddf)
+export LIBVMM=$(abspath ../../)
 
```

Comment:
At some point I want to move all the toolchain defines into one of two snippets: `gcc.mk` and `clang.mk`, but that's waiting for everything else

Reply:
Offline discussion: going to put this in another PR.